Create low-level SystemUtilities lib to resolve cyclic dependency between ErrorHandling and Parallel

nilsvu opened the pull request:
## Proposed changes

This PR moves `Parallel/Info.hpp`, `Parallel/Abort.hpp` and `Parallel/Exit.hpp` to `Utilities/System`. `Parallel/Printf.hpp` should also be moved there, but we have to clean up dependencies in `Utilities` before that can happen. By creating this low-level lib we can depend ErrorHandling on it so its cyclic dependency with `Parallel` is resolved.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
- If you are looking for `ErrorHandling/` headers, you'll find them in `Utilities/ErrorHandling/` now.
- If you are looking for low-level headers like `Parallel/Info.hpp`, see if you find what you need in `Utilities/System/`.
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe commented on src/ErrorHandling/Abort.cpp:
```diff
@@ -1,26 +1,15 @@
 // Distributed under the MIT License.
```
This commit seems to do more than a move

nilsvu replied:
reworded

nilsvu committed:
Add missing charm++ include in CharmCompatibility.cpp

This was previously implicitly included through Parallel and
ErrorHandling headers, but may not be anymore.

nilsvu commented:
Reworded commit message and added a missing include

nilsdeppe commented on src/ErrorHandling/AbortWithErrorMessage.cpp:
```diff
@@ -3,33 +3,32 @@
      << "############ ASSERT FAILED ############\n"
-     << "Node: " << Parallel::my_node() << " Proc: " << Parallel::my_proc()
-     << "\n"
+     << "Node: " << CkMyNode() << " Proc: " << CkMyPe() << "\n"
```
Since you now still basically depend on Parallel but have created a second library that does the same thing, why not just move the error handling to Parallel? Basically, this change makes it so two places explicitly depend on a bunch of Charm++ things, which I view as a regression. @kidder and I were thinking this cyclic dependency should be resolved by adding or refactoring `Parallel`.

nilsvu replied:
That means Parallel can't depend on anything that needs error handling, like Utilities, DataStructures, Informer, Options. Or those can't depend on error handling.

nilsdeppe replied:
Yes, this is where the "refactoring Parallel" comes in and why it hasn't been done yet :)

nilsvu replied:
I think in one way or another we need a lowest-level error handling library, and it will need to depend on (a subset of) Charm. We could move it to `Parallel/` but then we have to move `Breakpoint` and `Exceptions` elsewhere. So I was thinking we keep the `ErrorHandling` library and make it independent of the `Parallel` library but dependent on Charm (maybe a subset `Charmxx::error_handling` which I can add with #2680 as an imported target). `Parallel` presumably has to depend on the entirety of Charm, or on subsets when we factor it into sub-libs, e.g. for load-balancing.

nilsdeppe replied:
Sorry, I don't understand how moving `Breakpoint` and `Exceptions` causes an issue. Or even why they need to be moved. They could be moved but I don't understand your implied requirement

nilsvu replied:
Those have nothing to do with parallelization so shouldn't be in "Parallel", at least `Exceptions` surely doesn't

nilsdeppe replied:
I agree, they don't need to be there. But I still don't understand why they need to be with `ERROR`, etc., sorry

nilsvu replied:
They don't, so I said "We could move it to Parallel/ but then we have to move Breakpoint and Exceptions elsewhere". What I meant is we could move Breakpoint and Exceptions to, e.g., Utilities (though I'm not suggesting we do that).

nilsvu commented:
Notes from call on 2020-12-22 with @nilsdeppe and @kidder:

- The plan is to move `ErrorHandling/` to `Utilities/System` in a first PR (this one), inlining the Charm functions it needs. Then in separate PRs we move `Parallel/Info.hpp` to `Utilities/HardwareInfo/Info.hpp` (creating the even lower-level `HardwareInfo` library that `System` can depend on) and `Parallel/Printf.hpp` to `Utilities/System/Printf.hpp`.
- The `Utilities/` directory is for code that extents third-party or system / STL libraries. We should split the `Utilities` library into separate libs. Code that clearly extends a particular third-party lib could even be linked with that lib, e.g. headers in `Utilities/Blas` could be included in the `Blas` target.
- We should establish the rule that libraries can only depend on others that are on the same level or further down in the directory structure. For example, `DomainStructure` can depend on `Domain`, but not the other way round.

nilsleiffischer commented:
@nilsdeppe @kidder How about instead of moving `ErrorHandling` to `Utilities/System`, we just move it to `Utilities/ErrorHandling`? Then we can move `Parallel/Printf.hpp` to a new `Utilities/Logging/Printf.hpp`, i.e. make a low-level `Logging` library that `ErrorHandling` depends on.

nilsvu committed:
Move Abort.hpp to SystemUtilities and definition to source file

nilsvu committed:
Inline calls to Parallel::printf in ErrorHandling

This resolves the cyclic dependency between ErrorHandling and Parallel.
The plan is to move Parallel::printf to SystemUtilities, but we have to clean up
the dependencies in Utilities first.

nilsvu commented:
@nilsdeppe @kidder This is what I did now:

- Moved `Parallel/Abort.hpp -> Utilities/System/Abort.hpp`.
- Moved `Parallel/Exit.hpp -> Utilities/System/Exit.hpp`.
- Moved `Parallel/Info.hpp -> Utilities/System/ParallelInfo.hpp`.
- I kept `ErrorHandling` at top-level, because it's not really an extension of third-party / system / STL code, but our own. Do you agree or prefer to move it to `Utilities`?

The new `SystemUtilities` library should also be the home for `Parallel/Printf.hpp` once we have cleaned up the dependencies in `Utilities`.

nilsvu committed:
Move Parallel/Info.hpp to SystemUtilities

nilsvu committed:
Use sys::wall_time in Informer

nilsvu committed:
Move Parallel/Exit.hpp to SystemUtilities

nilsdeppe commented:
- I agree with the Abort and Exit moves, as well as the suggest walltime split.
- I thought on the call we said we wanted a `HardwareInfo` library for `Info.hpp` things
- I don't understand how error handling isn't an extension of the C++/C `abort`, `assert`, and `try-catch` exception systems. We literally have a function named `abort`, a macro named `ASSERT`, and an exception class in there. Things like breakpoint (and future stack trace capability) is also a direct extension/wrapper of POSIX-supplied functions.

nilsvu commented:
@nilsdeppe About `HardwareInfo`: I thought, why split it out of `SystemUtilities`, that's already as low-level as it gets. If you prefer I split it into a separate `HardwareInfo` lib, then which namespace would you suggest? And should `wall_time` remain in `SystemUtilities`?

nilsvu committed:
Move ErrorHandling to Utilities

nilsvu committed:
Fix some clang-tidy issues

markscheel commented:
My future PR to have mocked nodes and cores in ActionTesting will change Parallel/Info.hpp and essentially everything that depends on it, so it would be nice to have this merged soon so that I could rebase my stuff on this PR.

nilsvu commented:
Rebased, @markscheel could you do the second review then? @nilsdeppe I moved ErrorHandling to Utilities/ as you suggested, and see my comment above about your other questions.

markscheel commented on src/Utilities/System/Abort.cpp:
```diff
@@ -1,26 +1,16 @@
-#pragma once
+#include "Utilities/System/Abort.hpp"
 
 #include <charm++.h>
 #include <csignal>
 #include <cstdlib>
 #include <exception>
 #include <string>
```
Do we need all of those includes?  Shouldn't need `csignal` or `cstdlib` I think.

nilsvu replied:
Should be safe to delete them, done

markscheel reviewed (commented):
Looks good! One comment about possible extra includes.

markscheel commented:
About HardwareInfo - I'm fine with either Info.hpp stuff being in SystemUtilities or a new HardwareInfo library (maybe a `sys::hw_info` namespace?).  I plan to have two versions of all the Info.hpp functions in my PR:  the current version goes into a sys::charm namespace (or `sys::hw_info::charm` namespace) and a new version (templated on DistributedObject so that ActionTesting can modify cores/nodes) will be in the sys namespace (or `sys::hw_info`)

nilsvu commented:
@markscheel I think it would be good to discuss those plans briefly next Monday. If there's anything I can do in this PR that's uncontroversial and would help you out, please let me know. Anything that is not entirely clear at this point I'd like to defer to a follow-up PR.

markscheel reviewed (dismissed):
Everything here is fine with me. But @nilsdeppe might want a HardwareInfo library (which is also fine with me).

nilsvu commented:
rebased again to propagated changes to added code

nilsvu commented:
@markscheel could you take another look?

This PR touches a lot of files so a bunch of clang-tidy issues came up. I fixed the trivial ones, but I'd prefer to not touch the remaining ones (e.g. replacing C-style arrays).

markscheel reviewed (approved):
Looks good!

nilsdeppe commented:
Ignorring clang tidy, most are in boost root finders or should be NOLINT.
